### ISSUE TITLE
Notifications

### ISSUE THREAD

**Issue by naathyn** [naathyn/Ampersand#3]
This is a huge deal.  To be a social network, notifications are a must.  At the very least, there could be a notification model that holds all keys and values of when a user interacts with another user.  This would be anywhere from 'liking' a Micropost to commenting on a blog to replying to a message, etc.  Basically any interaction among two or more users needs a notification for the users involved.  (Much like Facebook or Twitter if you are familiar.)

I believe if we work together we can pull this off as it is a fairly large task.  

The next step would be to use ActionMailer to send out emails to the users involved in the interaction.

@JBekker, any ideas on how this model would be setup?  Are you thinking a polymorphic table to hold all these interactions and notifications?

**Comment by Jbekker** [comment]
A polymorphic table sounds good! I made a little proof of concept with a observer, observing the opinions table create notification records with a message.
OBSERVER:

``` ruby
class NotificationObserver < ActiveRecord::Observer
  observe Opinion, Blog, Comment

  def after_create(record)
    if record.class.name == "Opinion"
      micropost = record.like
      current_user = record.fan
      likes_user = micropost.user
      record.notifications.notify_opinion_create(current_user, likes_user, record)
    end
    if record.class.name == "Blog"
      #TODO
    end
    if record.class.name == "Comment"
      #TODO
    end
  end
end
```

TABLE:

``` ruby
class CreateNotifications < ActiveRecord::Migration
  def change
    create_table :notifications do |t|
      t.integer :user_id, :null => false
      t.text :message, :null => false
      t.integer :notify_id, :null => false
      t.string :notify_type, :null => false
      t.boolean :is_new, :default => true
      t.boolean :is_emailed, :default => false

      t.timestamps
    end
    add_index :notifications, :user_id
  end
end
```

NOTIFICATION MODEL:

``` ruby
class Notification < ActiveRecord::Base
  belongs_to :user
  belongs_to :notify, :polymorphic => true
  attr_accessible :user_id, :message, :notify_id, :notify_type, :is_emailed, :is_new

  def self.notify_opinion_create(current_user, likes_user, record)
    record.notifications.create(:user_id => likes_user.id, :message => "#{current_user.realname} likes your post")
  end

  def self.notify_blog_create(current_user, record)
    record.notifications.create(:user_id => current_user.id, :message => "#{current_user.realname} created a new blog named #{record.title}")
  end
end
```

**Comment by naathyn** [comment]
Looks great.  I am going to start trying to implement this asap.

**Comment by naathyn** [comment]
I am further implementing this but seems to have some problems as I have no notifications when I like a post.  I will continue to play with it.  I would really like to start thinking about how to get replies to a micropost attached to it so I can call the replies on the 'detach' button alongside the original micropost.

**Comment by Jbekker** [comment]
Did you add this line to your config/application.rb file? config.active_record.observers = :notification_observer

**Comment by naathyn** [comment]
Thanks, just found that a few minutes ago.  Works like a charm. For the views, I am basically thinking about new notifications dropping down from the navigation bar with a message + badge if new notifications.  Have you gone any further with this?

**Comment by Jbekker** [comment]
I didn’t have much time to work on it.
Whenever a user retrieves the notification for the first time, it should
update those records with is_new to false. Shouldn't be too hard! I'm also
thinking it's would be better to not store the full message but just the
parameters to build the message and use a view to build the message. That
way we could change the message texts and translate it etc..
What do you think?

**Comment by naathyn** [comment]
Yes I am truncating the notification's content and might show full notifications in an ordered list if a user wishes to view more than, say, 10 at a time.

**Comment by naathyn** [comment]
![notifications](https://f.cloud.github.com/assets/2221273/592730/49a86c20-ca23-11e2-9d69-62601715cfb8.png)

This is a naive implementation and still very basic, but just wanted to touch base with what kind of logic I am coming up with.
